### ISSUE TITLE
Problem with baseUrl and JavaScriptFetch, JavaScriptServiceWorkerRegistration...

### ISSUE THREAD

**Issue by kevincox** [assetgraph/assetgraph#1123]
Right now assets have a `baseUrl` which defaults to their `url`, or the `url` of the `baseUrl` of their nearest non-inline ancestor. However this is a flawed concept for JavaScript as it has a much more complex situation than that. Notably `import` is resolved relative to `baseUrl` however `fetch`, `navigator.serviceWorker.register` and similar are relative to the URL of the including document. Currently the code treats these as the same, which results in unfortunate results such as `navigator.serviceWorker.register('file:///home/kevincox/p/kevincox.ca/build/service-worker.js')`. 

However this problem is basically impossible unless assetgraph knows all possible references to the JavaScript file. There are cases where we can probably assume this such as when using `assetgraph-hashfiles` however there are also cases where this can not be assumed, such as if you are providing a JavaScript library that you expect users (and possibly your website as well) to access off of a well-known URL.

I encountered this problem with `serviceWorker.register` but I believe the same issue exists with `fetch`.

I think the best solution here is to not modify the referenced URL when the URL of the asset *containing* the reference is changed. There are a couple of cases where we can do something "clever" by handling the case where the only user of the script is moved, however this action at a distance is probably more confusing than it is helpful.

**Comment by kevincox** [comment]
Since I knew that all of my callers were coming from the domain of the target the following hack seems to be a viable workaround. A better solution would probably be remembering the original URL of the relation and always using that as the `baseUrl`.

```js
function fixJavaScriptDocumentRelativeRelations(Class) {
	Object.defineProperties(Class.prototype, {
		hrefType: {
			get() { return "rootRelative" },
			set() {},
		},
		baseUrl: {
			get() { return this.assetgraph ? this.assetgraph.root : this.url },
		},
	})
}
fixJavaScriptDocumentRelativeRelations(AssetGraph.JavaScriptFetch);
fixJavaScriptDocumentRelativeRelations(AssetGraph.JavaScriptServiceWorkerRegistration);
```

**Comment by papandreou** [comment]
Yeah, we used to have a really complex implementation that tried to get this right, but it still didn't always work due to the ambiguities you're pointng out. Would it be an option for you to use root-relative urls when `fetch`ing static assets and registering service workers?

**Comment by kevincox** [comment]
That is what my work around is doing. It seems like the "ideal" solution is fairly easy to define, but hard to implement:

- Top level users are all in the same directory: Relative URL from that directory.
- Top level users are all in the same domain: Root relative URL.
- Multiple domains: Absolute URL.

However actually identifying those conditions is incredibly difficult.

I think the best solution is to use the reference style used in the source. And changing the location of the asset shouldn't change the URL. (Changing the location of the target obviously will).

**Comment by papandreou** [comment]
> That is what my work around is doing

What I meant was -- would it be possible for you to use root-relative urls in _your_ source code when `fetch`ing static assets and registering service workers?

> I think the best solution is to use the reference style used in the source

That is already implemented -- we call it `hrefType` and it's maintained on each relation. If a reference is originally relative/root-relative/protocol-relative/absolute, we try to keep it like that when the source or target asset is moved. Much like in your breakdown, if the target asset of a relative reference is moved to another origin, it'll be changed to an absolute url. If the source asset is then also moved to the same origin, it snaps back to a relative url.

**Comment by kevincox** [comment]
The source code **does** have a root relative URL. I think the problem is that I am moving the file that registers the service worker to a different domain. This appears to be incorrectly triggering assetgraph to rewrite the URL to an absolute import because it concludes that you can't have a cross-domain `rootRelative` link. This assumption is wrong because even though the asset is on a different domain the `baseUri` **should** be the same domain. However assetgraph currently doesn't consider it to be that way.

**Comment by papandreou** [comment]
Hmm, that does sound like a bug. Any chance you could turn that into a test case?

**Comment by kevincox** [comment]
I'll see if I can extract it. The setup is fairly simple though.

- `index.html` has `<script src=sw.js>`
- `sw.js` has `navigator.serviceWorker.register("/service-worker.js");
- `graph.findAssets({path: "sw.js"}).forEach(a => a.url = "https://cdn.example");
- `graph.writeAssetsToDisc({protocol: "file:"}, "dist/")`

Completely untested but that is the gist of it.

**Comment by kevincox** [comment]
Ok, I made a test case. It is roughly as simple as expected 🙂

**Comment by kevincox** [comment]
Was this fix ever released?

**Comment by papandreou** [comment]
No, sorry, I forgot that 🙈 

I backported the fix in 220f93da4210b62686dae261a1362e3bcd1431ca and released 6.6.0 now.
